axelor-human-resource/src/main/java/com/axelor/apps/hr/web/timesheet/TimesheetController.java

```diff
@@ -594,4 +601,75 @@ public void removeAfterToDateTimesheetLines(ActionRequest request, ActionRespons
     }
     response.setValues(timesheet);
   }
+
+  public void initProjectPlanningTime(ActionRequest request, ActionResponse response) {
+    Timesheet timesheet = request.getContext().asType(Timesheet.class);
+
+    Employee employee = timesheet.getEmployee();
+    LocalDate fromDate = timesheet.getFromDate();
+    LocalDate toDate = timesheet.getToDate();
+
+    List<ProjectPlanningTime> projectPlanningTimeList =
+        Beans.get(ProjectPlanningTimeService.class)
+            .getProjectPlanningTimeIdList(employee, fromDate, toDate);
+
+    if (!ObjectUtils.isEmpty(projectPlanningTimeList)
+        && TimesheetRepository.STATUS_DRAFT == timesheet.getStatusSelect()) {
+      response.setValue("$projectPlanningTimeList", projectPlanningTimeList);
+      LocalDate todayDate = Beans.get(AppBaseService.class).getTodayDate(timesheet.getCompany());
+      response.setValue(
+          "$generationDate",
+          LocalDateHelper.isBetween(fromDate, toDate, todayDate) ? todayDate : null);
+
+      response.setAttr("projectPlanningTimePanel", "hidden", false);
+    } else {
+      response.setAttr("projectPlanningTimePanel", "hidden", true);
+    }
+  }
+
+  public void generateTimesheetLine(ActionRequest request, ActionResponse response)
+      throws AxelorException {
+    Timesheet timesheet = request.getContext().asType(Timesheet.class);
+    Object generationDateObject = request.getContext().get("generationDate");
+
+    if (generationDateObject == null) {
+      response.setError(I18n.get(HumanResourceExceptionMessage.NO_TIMESHEET_GENERATED_DATE));
+      return;
+    }
+    LocalDate generationDate = LocalDate.parse(generationDateObject.toString());
+    if (!LocalDateHelper.isBetween(
+        timesheet.getFromDate(), timesheet.getToDate(), generationDate)) {
+      response.setError(I18n.get(HumanResourceExceptionMessage.DATE_NOT_IN_TIMESHEET_PERIOD));
+      return;
+    }
+
+    List<Map<String, Object>> projectPlanningTimeList =
+        ((List<Map<String, Object>>) request.getContext().get("projectPlanningTimeList"))
+            .stream()
+                .filter(
+                    it ->
+                        Objects.nonNull(it.get("duration"))
+                            && (new BigDecimal(String.valueOf(it.get("duration")))).signum() > 0)
+                .collect(Collectors.toList());
+    if (!ObjectUtils.isEmpty(projectPlanningTimeList)) {
+      Beans.get(TimesheetLineCreateService.class)
+          .createTimesheetLinesUsingContextMap(projectPlanningTimeList, generationDate, timesheet);
```

Comment:
I think the code would become more maintainable if you converted in the controller the list of Map into a list ProjectPlanningTime.